databroker/eventsource/archiver.py

```diff
@@ -65,77 +107,285 @@ def descriptors_given_header(self, header):
+            datetime64 objects in UTC
+        timezone: str, optional
+            e.g., 'US/Eastern'
+        localize_times: bool, optional
+            If the times should be localized to the 'local' time zone.  If
+            True (the default) the time stamps are converted to the localtime zone.
+
+        Returns
+        -------
+        tables: dictionary of the pv pandas.DataFrames
+        """
+        
+        if timezone is None:
+            timezone = self.tz.zone
+
+        desc_uids = {}
+
         since, until = header['start']['time'], header['stop']['time']
-        for name, pv in six.iteritems(self.pvs):
-            _from = _munge_time(since, self.tz)
-            _to = _munge_time(until, self.tz)
-            params = {'pv': pv, 'from': _from, 'to': _to}
-            req = requests.get(self.archiver_addr, params=params, stream=True)
-            req.raise_for_status()
-            raw, = req.json()
-            timestamps = [x['secs'] for x in raw['data']]
-            data = [x['val'] for x in raw['data']]
-            for seq_num, (d, t) in enumerate(zip(data, timestamps)):
-                doc = {'data': {name: d},
-                       'timestamps': {name: t},
-                       'time': t,
-                       'uid': 'ephemeral-' + str(uuid.uuid4()),
-                       'seq_num': seq_num,
-                       'descriptor': desc_uids[pv]}
-                yield doc               
-
-    def table_given_header(self, header, fields=None):
-        # TODO: Add timezone goodies
-        no_fields_filter = False
-        if fields is None:
-            no_fields_filter = True
-            fields = []
-        fields = set(fields)
-        docs = list(self.docs_given_header(header=header,
-                                    fill=False, fields=fields))
-        return pd.DataFrame.from_records(data=docs, index='seq_num')
-
-    def fill_event(self, *args, **kwrags):
-        raise NotImplementedError()
-
-    def fill_table(self, *args, **kwargs):
-        raise NotImplementedError()
+  
+        df = pd.DataFrame()
 
+        if stream_name in self.stream_names_given_header(header):
+            name = stream_name[9:] # archiver_<name>
+            pv = self.pvs[name]
+            df = self._table_given_times(pv, since, until)
+            df.rename(columns = {"time": "time", "data" : stream_name}, inplace=True)
+            new_index = np.arange(1, df.index.size +1)
+            df.index = new_index
 
+        return df
+ 
 def _munge_time(t, timezone):
     """Close your eyes and trust @arkilic
-
     Parameters
     ----------
     t : float
         POSIX (seconds since 1970)
     timezone : pytz object
         e.g. ``pytz.timezone('US/Eastern')``
-
     Return
     ------
     time
         as ISO-8601 format
     """
     t = datetime.fromtimestamp(t)
     return timezone.localize(t).replace(microsecond=0).isoformat()
+
+class ArchiverReader(object):
```

Comment:
Can this be merged into https://github.com/NSLS-II/databroker/blob/master/databroker/_core.py#L2086 ?  I don't think we need a separate configuration system for this.

Reply:
yes, it is used as (1) a factory for ArchiverEventSources:
https://github.com/malitsky/nsls2/blob/master/areventsource/notebooks/csx_with_archivers.yml

and as (2) a stand-alone class for the CSX extensions
https://github.com/malitsky/nsls2/blob/master/areventsource/notebooks/archiver-reader.db.5.13.18.ipynb

Reply:
As discussed in person, please make this class private.